README
Data::BitStream version 0.07

A Moo class providing read/write access to bit streams.  This includes many
integer coding methods as well as straightforward ways to implement new codes.

Bit streams are often used in data compression and in embedded products where
memory is at a premium.  Using variable length codes allows high performance
compression of integer data.

The Data::BitStream::XS module is also available, and is a near drop-in
replacement for this module, though it does not use Moo.  The Data::BitStream
module will internally use the XS module if it is found, giving a lot of the
performance benefit without any other changes necessary.


SYNOPSIS

  use Data::BitStream;
  my $stream = Data::BitStream->new;
  $stream->put_gamma($_) for 1..20;
  $stream->rewind_for_read;
  my @values = $stream->get_gamma(-1);

See the POD module documentation for examples and more information on all the
methods included.  There are also examples and benchmarks in the examples
directory in the distribution, including a functional lossless image codec.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


DEPENDENCIES

This module requires these other modules and libraries:

  Test::More
  Moo
  Moo::Role
  MooX::Types::MooseLike

It works identically with Moo, Mouse, or Moose.  Moo with Class::XSAccessor
installed is as fast as Mouse for some operations, but not all.  If speed
is important, install Data::BitStream::XS.  If speed is more important than
extensibility, use Data::BitStream::XS directly.

I have tested this using perl versions 5.6.2 to 5.20.0.

It works with Perl 5.6.2, with a caveat.  On 32-bit O/S, 64-bit Perl can have
broken arithmetic (two unequal 64-bit numbers will compare as equal).  We force
maxbits to 32 if it looks like it won't work with 64.


COPYRIGHT AND LICENCE

Copyright (C) 2011-2013 by Dana Jacobsen <dana@acm.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.14.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Data::BitStream.

0.09    ?
        - Update CPAN meta data

0.08    16 Aug 2014
        - Updates to documentation.
        - Goldbach encoding almost 2x faster (sped up pair search).

0.07    1 Jan 2013
	- Change from Mouse to Moo.  Mouse has great performance, but the Perl
	  community is moving toward Moo/Moose and away from Any:Moose and
	  Mouse.  In most cases Moo will be similar performance assuming the
	  Class::XSAccessor module is installed.
	- Remove some Unicode characters that snuck into references.
	- Much faster primes for big Goldbach codes.  DBXS is still ~30x
	  faster overall, but almost all of that is the pair search.
        - Add BER and Varint codes.

0.06   28 May 2012
	- Make all tests numbered.
	- Documentation additions (Pod Coverage testing).
	- get_arice / put_arice modify the k parameter, just like XS code.
	- Update Additive / Goldbach to reflect DBXS and MPFS changes.

0.05   17 May 2012
	- Make a couple 64-bit tests work properly with the 'xs' module on
	  LLP64 systems.
	- Add put_raw, from_raw uses it.
	- String has custom to_raw and put_raw.
	- Inline C for decorrelator / predictor in image compression example.
	- Better behavior when doing bad reads.
	- Unify a lot of error strings.
	- Add generalized Fibonacci coding (order 2-16).
	- Add additive and Goldbach G1/G2 codes from Fenwick's 2002 IEEE paper.
	- Add comma codes (including ternary comma codes e.g. Fenwick 1996).
	- Add block-based taboo codes (Pigeon 2001).

0.04    9 Nov 2011
	- Fix late regression in one of the tests, unpack "W" -> "C".
	- Minor test and documentation changes.
	- Make most error strings consistent.

0.03   Mon Nov  7 06:00:00 2011
	- More work on unit tests.
	- Add BLVec class to use XS.  Inefficient but it works (and even with
	  the silly setpos/setlen after every call, it's much faster than PP).
	- Don't use Vec and BitVec implementations unless specifically asked.
	- Lossless image codec example.
	- Simple file support (read / write).
	- Text methods (code_get, code_put, add_code)
	- Performance.
	- Clarify POD compression example slightly.
	- More tests on 32-bit and big-endian machines.

0.02  Fri Sep 30 08:16:13 2011
	- Create dist with tool instead of by hand.  Uses version number.
	- WordVec was setting the Vec variables.

0.01  Mon Sep 26 10:12:53 2011
	- original version; created by h2xs 1.23 with options
		-XAn Data::BitStream

	- 24 Sep 2011, add Boldi-Vigna Zeta, Levenstein Codes, and Baer codes
	- 19 Sep 2011, make Golomb take a sub, better default impls
	- 14 Sep 2011, add skip(), optimizations
	- 13 Sep 2011, more tests, harden the reader/writer distinction
	- 12 Sep 2011, add Delta and Omega
	- 11 Sep 2011, add ExpGolomb, and fix readahead
	- 9 Sep 2011, split into Base/impls/codes, first unit tests,
		add Fibonacci, Golomb, and Rice encoding.
	- 8 Sep 2011, Vec and BitVec code, with Gamma and GG3
	- 24 Aug 2011, first Perl version, Unary


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile1(
    NAME         => 'Data::BitStream',
    ABSTRACT     => "A class implementing a stream of bits and coding methods",
    VERSION_FROM => 'lib/Data/BitStream.pm', # finds $VERSION
    LICENSE      => 'perl',
    AUTHOR       => 'Dana A Jacobsen <dana@acm.org>',

    TEST_REQUIRES => {
                     # I'm using subtest.  Could go back to 0.82 without it.
                     'Test::More'       => '0.94',
                     },
    PREREQ_PM => {
                     # Using just the import method of Exporter
                     'Exporter'         => '5.57',
                     'Carp'             => 0,

                     # This module used to use Mouse:
                     # 'Mouse'            => '0.58',
                     # 'Mouse::Role'      => '0.58',
                     # Then we tied Moose (2.0205 - 2.0602):
                     # 'Moose'            => 0,
                     # 'Moose::Role'      => 0,
                     # 'MooseX::Types::Moose' => 0,
                     # And now Moo.
                      'Moo'                    => 1.000006,
                      'Moo::Role'              => 0,
                      'MooX::Types::MooseLike' => 0,
                     # Technically not needed, but big performance help
                      'Class::XSAccessor'      => 0,
                 },

    META_MERGE => {
               'meta-spec'  => {
                  version => '2',
                  url     => 'http://search.cpan.org/perldoc?CPAN::Meta::Spec',
               },
               resources  => {
                 license    => [ 'http://dev.perl.org/licenses/' ],
                 homepage   => 'https://github.com/danaj/BitStream',
                 repository => {
                   url  => 'https://github.com/danaj/BitStream',
                 },
               },
               prereqs => {
                 runtime => {
                   recommends => {
                     'Data::BitStream::XS' => 0.06,
                     'Math::Prime::Util'   => 0,    # For Goldbach codes
                   },
                 },
               },
             },

    # I have not tested with any version of Perl previous to 5.6.2.  v5.6.2
    # does work, though is forced to 32-bit if it detects 64-bit arithmetic is
    # broken.  Data::BitStream::XS works fine on 5.6.2 because it is using C
    # for the bit operations.
    MIN_PERL_VERSION => 5.006002,

);

sub WriteMakefile1 { 
  my %params = @_;
  my $eumm_version = $ExtUtils::MakeMaker::VERSION;
  $eumm_version = eval $eumm_version;

  if ($params{TEST_REQUIRES} and $eumm_version < 6.6303) {
      $params{BUILD_REQUIRES}={ %{$params{BUILD_REQUIRES} || {}} , %{$params{TEST_REQUIRES}} };
      delete $params{TEST_REQUIRES};
  }
  if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
      #EUMM 6.5502 has problems with BUILD_REQUIRES
      $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
      delete $params{BUILD_REQUIRES};
  }
  delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
  delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
  delete $params{META_MERGE} if $eumm_version < 6.46;
  delete $params{META_ADD} if $eumm_version < 6.46;
  delete $params{LICENSE} if $eumm_version < 6.31;
  delete $params{AUTHOR} if $] < 5.005;
  delete $params{ABSTRACT_FROM} if $] < 5.005;
  delete $params{BINARY_LOCATION} if $] < 5.005;

  WriteMakefile(%params);
}
